trino/trino_test.go

```diff
@@ -249,6 +250,99 @@ func TestQueryForUsername(t *testing.T) {
 	}
 }
 
+type TestQueryProgressCallback struct {
+	statusMap map[time.Time]string
+}
+
+func (qpc *TestQueryProgressCallback) Update(qr QueryProgressCallbackResponse) {
+	qpc.statusMap[time.Now()] = qr.QueryStats.State
+}
+
+func TestQueryProgressWithCallback(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping test in short mode.")
+	}
+	c := &Config{
+		ServerURI:         *integrationServerFlag,
+		SessionProperties: map[string]string{"query_priority": "1"},
+	}
+
+	dsn, err := c.FormatDSN()
+	require.NoError(t, err)
+
+	db, err := sql.Open("trino", dsn)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		assert.NoError(t, db.Close())
+	})
+
+	callback := &TestQueryProgressCallback{}
+
+	_, err = db.Query("SELECT 2", sql.Named("X-Trino-Progress-Callback", callback))
+	assert.EqualError(t, err, ErrInvalidProgressCallbackHeader.Error(), "unexpected error")
+}
+
+func TestQueryProgressWithCallbackPeriod(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping test in short mode.")
+	}
+	c := &Config{
+		ServerURI:         *integrationServerFlag,
+		SessionProperties: map[string]string{"query_priority": "1"},
+	}
+
+	dsn, err := c.FormatDSN()
+	require.NoError(t, err)
+
+	db, err := sql.Open("trino", dsn)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		assert.NoError(t, db.Close())
+	})
+
+	statusMap := make(map[time.Time]string)
+	callback := &TestQueryProgressCallback{
+		statusMap: statusMap,
+	}
+	callbackPeriod, err := time.ParseDuration("1ms")
+
+	rows, err := db.Query("SELECT 2",
+		sql.Named("X-Trino-Progress-Callback", callback),
+		sql.Named("X-Trino-Progress-Callback-Period", callbackPeriod),
+	)
+	require.NoError(t, err, "Failed executing query")
+	assert.NotNil(t, rows)
+
+	defer rows.Close()
+	for rows.Next() {
+		var ts string
+		require.NoError(t, rows.Scan(&ts), "Failed scanning query result")
+
+		assert.Equal(t, "2", ts, "Expected value does not equal result value")
+	}
+
+	if err = rows.Err(); err != nil {
+		t.Fatal(err)
+	}
+
+	// sort time in order to calculate interval
+	var keys []time.Time
+	for k := range statusMap {
```

Comment:
Tests are failing now with a race condition here. The race detector sees two goroutines that read and write from the same map and it cannot determine if the goroute that calls the update callback has finished or not. To eliminate this possible race, you could close the `statsCh` channel before exiting the goroutine and try to read from it after closing `doneCh` in `ExecContext` and `QueryContext` to make sure it has been closed.

Reply:
While I was trying to close the `statsCh` channel inside the go routine, I noticed that `rows.Next()` calls fetch function, which calls updateCallback function.
So I added code to check and return if `statsCh` channel is close. Did I understood your intention correctly?
- https://github.com/trinodb/trino-go-client/pull/14/commits/f2af3433d9581a7cd1b744754b5ff8845ea4cb52#diff-0142aac4be66432bab473dd519af9593a22e6842a9a15a5251c14d97b3208519R791

Reply:
That should not be necessary, but it means that `doneCh` is closed prematurely. I overlooked this, it should be closed in `func (qr *driverRows) Close() error`. `Next()` should never be called after `Close()`.

Reply:
How about this? calling done channel to close go routine callback channels
```go
func (qr *driverRows) Close() error {
	if qr.stmt.conn.progressUpdater != nil {
		qr.doneCh <- struct{}{}
	}
```
with go function as 
```go
		go func() {
			for {
				select {
				case stats := <-st.statsCh:
					st.conn.progressUpdater.Update(stats)
				case <-st.doneCh:
					close(st.statsCh)
					close(st.doneCh)
					return
				}
			}
		}()

```

Reply:
You can't close a channel twice. If the `case <-st.doneCh` branch executes, this channel should be already closed. And you can't receive anything on an empty struct channel (because an empty struct has a size of 0, so it's not even sent to the channel).

Reply:
So the `doneCh` channel should be closed in the `rows.Close()` method. It would be best if all channels were uni-directional, and closing a channel is equal to writing to it, so you shouldn't close it where you read from it because that would mean it has to be bi-directional. Closing `statsCh` in the goroutine is an exception, we could avoid that by introducing yet another channel, but I don't think we have to right now.

Reply:
I don't quite understand when you mean by `And you can't receive anything on an empty struct channel (because an empty struct has a size of 0, so it's not even sent to the channel).` because on playground, when calling `close(done)` in line 27 or calling `done <- struct{}{}` in line 26 results the same and sending to empty struct channel works.
- https://go.dev/play/p/L91TX9ngoNI

If I simply remove `close(st.doneCh)`, i do not get any errors, `however`, I get race condition error if i move `close(st.doneCh)` and replace it at `qr.doneCh <- struct{}{}`
![image](https://user-images.githubusercontent.com/25147023/185592196-0ec1e63d-578e-47ea-8049-37f6910c1e6d.png)

```
==================
WARNING: DATA RACE
Read at 0x00c00028eba0 by goroutine 46:
  runtime.mapdelete()
      /usr/local/go/src/runtime/map.go:685 +0x49c
  github.com/trinodb/trino-go-client/trino.TestQueryProgressWithCallbackPeriod()
      /Users/user/work/trino-go-client/trino/trino_test.go:332 +0x7f0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x198

Previous write at 0x00c00028eba0 by goroutine 47:
  runtime.mapaccessK()
      /usr/local/go/src/runtime/map.go:511 +0x1ec
  github.com/trinodb/trino-go-client/trino.(*TestQueryProgressCallback).Update()
      /Users/user/work/trino-go-client/trino/trino_test.go:258 +0x84
  github.com/trinodb/trino-go-client/trino.(*driverStmt).exec.func1()
      /Users/user/work/trino-go-client/trino/trino.go:786 +0xc4

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x5b8
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0xac
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x198
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x780
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x928
  github.com/trinodb/trino-go-client/trino.TestMain()
      /Users/user/work/trino-go-client/trino/integration_test.go:101 +0x95c
  main.main()
      _testmain.go:129 +0x290

Goroutine 47 (finished) created at:
  github.com/trinodb/trino-go-client/trino.(*driverStmt).exec()
      /Users/user/work/trino-go-client/trino/trino.go:782 +0x614
  github.com/trinodb/trino-go-client/trino.(*driverStmt).QueryContext()
      /Users/user/work/trino-go-client/trino/trino.go:693 +0x58
  database/sql.ctxDriverStmtQuery()
      /usr/local/go/src/database/sql/ctxutil.go:82 +0x1dc
  database/sql.rowsiFromStatement()
      /usr/local/go/src/database/sql/sql.go:2794 +0x168
  database/sql.(*DB).queryDC()
      /usr/local/go/src/database/sql/sql.go:1752 +0x3e4
  database/sql.(*DB).query()
      /usr/local/go/src/database/sql/sql.go:1700 +0x124
  database/sql.(*DB).QueryContext()
      /usr/local/go/src/database/sql/sql.go:1674 +0xa4
  database/sql.(*DB).Query()
      /usr/local/go/src/database/sql/sql.go:1691 +0x4e4
  github.com/trinodb/trino-go-client/trino.TestQueryProgressWithCallbackPeriod()
      /Users/user/work/trino-go-client/trino/trino_test.go:311 +0x438
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x198
==================
    testing.go:1152: race detected during execution of test
```

Reply:
You're right. I wrongly remembered how it works. No surprises here: https://go.dev/play/p/KxyNdF1TiUp

But my point is valid though, that we don't need to send any signals over the empty channel and it's enough to close it once when no more progress updates will be made.

Reply:
>  we don't need to send any signals over the empty channel and it's enough to close it once when no more progress updates will be made.

Hmm.. I wonder why I `get race error` if I call close(qr.doneCh) at Close() and `no error` happens if I do `qr.doneCh <- struct{}{}` instead? 
Could it be because close() doc saids `has the effect of shutting down the channel after the last sent value is received` so doneCh waits for last sent value before closing?

Reply:
Can you push the changes where you get the race?

Reply:
@nineinchnick here made changes

Reply:
Here's a full diff:
[test.txt](https://github.com/trinodb/trino-go-client/files/9382797/test.txt)

Reply:
Regarding this code, isn't this used for channel synchronization ?  I was reluctant on using this because it will wait until statsCh has closed.
```go
 if st.statsCh != nil {
               <-st.statsCh
       }
```

Reply:
> I previously suggested doing this in the driverRows.Close(), but it's not aware of the statsCh so it needs to be done in the driverStmt.Close()

I first thought that it would be more appropriate to close channel `AFTER rows iterator` finishes, as it also calls FETCH function. (makes sense because data is currently at FETCH status)
But, when I debugged, I noticed that `func (st *driverStmt) Close` function is called `after` `func (qr *driverRows) Close() error {`finishes, so it would further makes sense to finish when statement is done.

Reply:
If it is allright, then I'll squash commits 
Also, I added comment to explain what driverStmt close does in [Sql.go](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/database/sql/sql.go;l=3330)
```
// Close closes statement just before releasing connection
func (st *driverStmt) Close() error {
```
- https://github.com/trinodb/trino-go-client/pull/14/commits/30d49d1238896c00c5001754d3e8b9dbab0740df